lib/schedule/minischedule/load.ex

```diff
@@ -0,0 +1,157 @@
+defmodule Schedule.Minischedule.Load do
+  @moduledoc """
+  Load the HASTUS data into minischedules data
+  """
+
+  alias Schedule.Helpers
+  alias Schedule.Hastus.Activity
+  alias Schedule.Hastus.Trip
+  alias Schedule.Minischedule.Block
+  alias Schedule.Minischedule.Break
+  alias Schedule.Minischedule.Piece
+  alias Schedule.Minischedule.Run
+
+  @spec from_hastus([Activity.t()], [Trip.t()]) ::
+          %{runs: Run.by_id(), blocks: Block.by_id()}
+  def from_hastus(activities, trips) do
+    activities_by_run = group_activities_by_run(activities)
+    trips_by_run = group_trips_by_run(trips)
+    run_groups = pair_activities_and_trips(activities_by_run, trips_by_run)
+    runs_and_pieces = Enum.map(run_groups, &run_and_pieces_from_run_group/1)
+    runs_by_id = Map.new(runs_and_pieces, fn {run_key, run, _pieces} -> {run_key, run} end)
+    pieces = Enum.flat_map(runs_and_pieces, fn {_run_key, _run, pieces} -> pieces end)
+    blocks_by_id = blocks_from_pieces(pieces)
+
+    %{
+      runs: runs_by_id,
+      blocks: blocks_by_id
+    }
+  end
+
+  # All the activities on a run
+  @typep activity_group :: {Run.key(), [Activity.t()]}
+
+  @spec group_activities_by_run([Activity.t()]) :: [activity_group()]
+  defp group_activities_by_run(activities) do
+    Helpers.split_by(activities, &run_key_for_activity/1)
+  end
+
+  @spec run_key_for_activity(Activity.t()) :: Run.key()
+  defp run_key_for_activity(activity) do
+    {activity.schedule_id, activity.run_id}
+  end
+
+  # The trips that form a piece together
+  @typep trip_group :: {Run.key(), [Trip.t()]}
+
+  @spec group_trips_by_run([Trip.t()]) :: [trip_group()]
+  defp group_trips_by_run(trips) do
+    Helpers.split_by(trips, &run_key_for_trip/1)
+  end
+
+  @spec run_key_for_trip(Trip.t()) :: Run.key()
+  defp run_key_for_trip(trip) do
+    {trip.schedule_id, trip.run_id}
+  end
+
+  @typep run_group :: {Run.key(), [Activity.t()], [Trip.t()]}
+
+  # Assumes inputs are both sorted by run_key.
+  # Assumes no runs are in trips.csv but not in activities.csv
+  @spec pair_activities_and_trips([activity_group()], [trip_group()]) :: [run_group()]
+  defp pair_activities_and_trips([], _trip_groups) do
+    []
+  end
+
+  defp pair_activities_and_trips([activity_group | other_activity_groups], trip_groups) do
+    {run_key, activities} = activity_group
+
+    case trip_groups do
+      [{^run_key, trips} | other_trip_groups] ->
```

Comment:
If `activity_groups` and `trip_groups` were maps instead of lists of tuples would a lot of this complexity and recursion go away? I.e. could you just iterate over activities if you could reference trips by `run_key`?

(If so I might retract the unit test comment.)

Reply:
Yeah, it all became a lot better. 👍